### PR TITLE
Add AbstractImageTeacher and multiprocess downloader

### PR DESCRIPTION
**Pull request by mwillwork**
Various fixes for AbstractImageTeacher. 

Example commands for training in accompanying PR.

### CODE REVIEW

**Comment by stephenroller** [comment]
Bump

File: parlai/core/build_data.py [hunk 1]
```diff
@@ -328,6 +328,9 @@ def download_multiprocess(
     """
     Download items in parallel (e.g. for an image + dialogue task)
 
+    Note: "of threading, multiprocess and pytorch.multiprocessing pick two".
+    These three don't all play well together. On OS X, may hang upon successful finish.
```
[the remark this answers is not in the document]
Comment (mwillwork) [reply parlai/core/build_data.py]: Hmm I don't think we really should be passing in opt to this function. How else could we accomplish that?

File: parlai/core/teachers.py [hunk 2]
```diff
@@ -1454,6 +1460,16 @@ def get_image_mode_features_path(self, task, image_model_name, dt):
...
 
+    def is_image_mode_buildable(self, model_name):
+        """
+        Is buildable if features can be calculated by ImageLoader.
+
+        Users may wish to compute features for the dataset offline and use in
+        the model, in which case, the image model should return False and
+        get_image_features() should be overriden in subclass.
+        """
+        return model_name in ImageLoader.get_available_model_names()
```
[the remark this answers is not in the document]
Comment (mwillwork) [reply parlai/core/teachers.py]: Ah no - it still should be ImageLoader.

File: parlai/core/teachers.py [hunk 3]
```diff
@@ -1600,40 +1635,43 @@ def next_example(self):
...
+        if self.include_image:
+            if self.image_features_dict is not None:
+                # We have specified an image model and it uses image features either
+                # generated by ImageLoader or by pre-computation and manually
+                # placed on path
+                if self.example is None:
+                    # Call to super method calls get() to get next example
+                    self.example, self.imageEpochDone = super().next_example()
+                return_result = self.example, self.imageEpochDone
+            elif self.include_image:
```
[the remark this answers is not in the document]
Comment (mwillwork) [reply parlai/core/teachers.py]: Agreed

File: parlai/core/build_data.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (stephenroller) [reply parlai/core/build_data.py]: Hmm, maybe it's the same as #2010?

[the remark this answers is not in the document]
Comment (stephenroller) [reply parlai/core/build_data.py]: Could also have this method revert to single process if it's running on OS X 🙈

[the remark this answers is not in the document]
Comment (mwillwork) [reply parlai/core/build_data.py]: Perhaps! (Re: 2010!)

[the remark this answers is not in the document]
Comment (mwillwork) [reply parlai/core/build_data.py]: Not going to add the feature right now of reverting to single process if it's OS X. Let me know if you really think it's necessary.

File: parlai/core/teachers.py [hunk 4]
```diff
@@ -1498,16 +1515,32 @@ def setup_image_features(self, data_path):
...
             if self.is_image_mode_buildable(self.image_mode):
+                # TODO: Awkward to modify the input opt but needed to use
+                # TODO: ImageLoader functionality. Is from comment_battle,
+                # TODO: will refactor this at some point soon most likely
+                image_loader_opt = self.opt.copy()
+                image_loader_opt['image_mode'] = (
+                    self.image_mode if self.include_image else 'none'
+                )
+
+                image_loader_opt['image_size'] = 256
```
[the remark this answers is not in the document]
Comment (mwillwork) [reply parlai/core/teachers.py]: We won't get here if image_mode is none (b/c model won't be buildable), so I don't think I need to check. But, you are right - those fields aren't in the opt if image_mode is none.

**Pull request merged**